cmd/lakectl/cmd/fs.go

```diff
@@ -121,11 +122,26 @@ var fsCatCmd = &cobra.Command{
 	Run: func(cmd *cobra.Command, args []string) {
 		client := getClient()
 		pathURI := uri.Must(uri.Parse(args[0]))
-		resp, err := client.GetObjectWithResponse(cmd.Context(), pathURI.Repository, pathURI.Ref, &api.GetObjectParams{
-			Path: *pathURI.Path,
-		})
-		DieOnResponseError(resp, err)
-		Fmt("%s\n", string(resp.Body))
+		direct, _ := cmd.Flags().GetBool("direct")
+		var contents []byte
+		if direct {
+			_, body, err := helpers.ClientDownload(cmd.Context(), client, pathURI.Repository, pathURI.Ref, *pathURI.Path)
+			if err != nil {
+				DieErr(err)
+			}
+			defer body.Close()
+			contents, err = ioutil.ReadAll(body)
+			if err != nil {
+				DieErr(err)
+			}
+		} else {
+			resp, err := client.GetObjectWithResponse(cmd.Context(), pathURI.Repository, pathURI.Ref, &api.GetObjectParams{
+				Path: *pathURI.Path,
+			})
+			DieOnResponseError(resp, err)
+			contents = resp.Body
+		}
```

Comment:
I think a helper that chooses the right method to download is appropriate here as this will probably be repeated elsewhere.

Reply:
Yes, but not yet: don't want to break any workflows (yet).  Or (even better) all adapters will support it.  We will still need indirect uploads (when no credentials exist), but it should be less common.
The thing is, right now it takes at least one extra RPC to each of lakeFS and S3 to determine this automatically.
Opened #1726 to track.